### ISSUE TITLE
Cannot open more than one session in the same browser

### ISSUE THREAD

**Issue by gormux** [aiden0z/guacamole-auth-jwt#6]
Hello,

I use this auth plugin to make a custom portal.
I can open a session to a computer, no problem.
But if I open a second session, the first one disconnects.

I use the "normal" guacamole client : 
http://my_guacamole:8081/guacamole?token=%s&&GUAC_DATA_SOURCE=jwt&GUAC_ID=%s&GUAC_TYPE=c&GUAC_USERNAME=%s
where token is obtained via /guacamole/api/token
GUAC_ID is randomly generated, and GUAC_USERNAME obtained from the token answer of the API.

Using a very basic client (this one : https://github.com/aiden0z/guacamole-auth-jwt/issues/1), I don't have the issue.

So I suppose the "normal" guacamole client does not like something in the way I use it, but I cannot find what. Maybe it does not support using different usernames in the same browser ?

I can open multiple connections using multiple browsers, so I'm pretty sure the guacamole client is the cause.

Is this a known issue ? Are there other people using this auth plugin and the normal guacamole client ?

**Comment by aronson** [comment]
We have JWT integration at my company using this plugin. I've noticed that the normal guacamole client will not support multiple "authorized" sessions in the same browser session.

I think it's related to how Guacamole caches credential tokens, and that when a new session is open it somehow overrides the old credential. It's related to session state as far as I can tell.

Hopefully you find a solution!

**Comment by gormux** [comment]
So, you use another client ? If so, is it available to the public, or is it internal ?

**Comment by iota-008** [comment]
Hi! can we pass the authToken in Headers instead of URL?

**Comment by aiden0z** [comment]
@iota-008  yes, the new v1.5.4 version has supported pass token in headers, please refer the [readme](https://github.com/aiden0z/guacamole-auth-jwt?tab=readme-ov-file#version-154) for details.

**Comment by Ed1lan** [comment]
@gormux I am not using this extension but I ended up here with the same problem, so, I'll say what I discovered if that helps anyone in the future.
Guacamole itself handles multiple sessions in the same browser because it uses the same API token in each of them. Whenever a session is opened, the access credentials are stored in LocalStorage and overrides previous ones on the process, that's probably why multilpe sessions are not working for you either for me, you must use a single token access in the same browser to fix the issue.
Sorry if my english is bad, hope it helps someone in the future.

**Comment by swethacheruku1** [comment]
@gormux could you please let me know whether this issue is resolved or not 
if it is resolved let me know the solution for this 
i can briefly explain what am doing am creating connections using the guacamole rest api endpoints and the token will be different for each connection but it corresponds to the same user only when i try to open multiple sessions inital ones are asking for login latest is working as expected this is emergency solution requiresd for us i hope i find solution as soon as possible

**Comment by KKKKKKevinGor** [comment]
Sorry if my english is bad, hope it helps someone in the future.

Apache Guacamole 同一浏览器多开 RDP 连接方案
问题描述
在同一个 Chrome 浏览器中打开多个 Guacamole RDP 连接 URL 时，新打开的 tab 会自动跳转到旧的连接，无法同时访问多个不同的 RDP 主机。

问题原因
Guacamole 前端使用 localStorage 存储认证信息（GUAC_AUTH token），而同一浏览器的所有 tab 共享同一个 localStorage。当打开新的连接 URL 时：

新 token 覆盖 localStorage 中的旧 token
旧 tab 的认证信息失效
Guacamole 前端检测到已有 session，将新 tab 重定向到旧连接
解决方案
通过 OpenResty (Nginx) 的 Lua 脚本注入 JavaScript，将 Guacamole 的 localStorage 操作重定向到 sessionStorage，实现每个浏览器 tab 拥有独立的 session 存储。

1. 修改 OpenResty/Nginx 配置
在 location ^~ /guacamole/ 块中添加 Lua 脚本注入：

location ^~ /guacamole/ {
    proxy_pass http://10.0.0.15:8080;
    # ... 其他 proxy 配置 ...
    
    # 关闭上游 gzip，确保可以修改响应内容
    proxy_set_header Accept-Encoding "";
    
    # 清除 Content-Length，因为注入会改变响应大小
    header_filter_by_lua_block {
        ngx.header.content_length = nil
    }
    
    # 注入 JS 脚本
    body_filter_by_lua_block {
        local chunk = ngx.arg[1]
        if chunk then
            local inject_script = [[<script>...</script></head>]]
            ngx.arg[1] = chunk:gsub("</head>", inject_script)
        end
    }
}
2. 注入的 JavaScript 核心逻辑
// 为每个 tab 生成唯一 ID，存储在 sessionStorage（tab 独立）
var tabId = sessionStorage.getItem('__guacTabId');
if(!tabId){
    tabId = 'tab_' + Math.random().toString(36).substr(2,9) + '_' + Date.now();
    sessionStorage.setItem('__guacTabId', tabId);
}

// 拦截 localStorage，将 GUAC 相关的 key 加上 tabId 前缀存到 sessionStorage
Object.defineProperty(window, 'localStorage', {
    get: function() { return fakeLocalStorage; }
});
3. API 端确保每个请求生成唯一标识
def generate_token(self, hostname, username, password, ...):
    session_uuid = str(uuid.uuid4())[:8]
    guac_username = f"rdp_{session_uuid}"  # 唯一用户名
    connection_name = f"{hostname}_{session_uuid}"  # 唯一连接名
技术要点
组件	作用
sessionStorage	每个浏览器 tab 独立，关闭 tab 后清除
localStorage	同一域名下所有 tab 共享
tabId	为每个 tab 生成唯一标识，区分不同 session
Lua body_filter	在 Nginx 层修改 HTML 响应，注入 JS
最终效果
同一 Chrome 浏览器可同时打开多个不同 RDP 主机的 tab
每个 tab 的 session 完全独立，互不干扰
无需用户安装多个浏览器或使用无痕模式

**Comment by betaXpolaris** [comment]
hope this also help to someone is facing this issue. 

in the guacamole deployment you need to do this: 

editing the app/storage/services/localStorageService.js :

you need to edit all the localstorage.* to -> windows.sessionstorage / sessionStorage

And in index.html  add this <script>try{sessionStorage.removeItem("GUAC_AUTH_TOKEN")}catch(e){}</script> in <head> tag to force a clean start even across duplicated or newly opened tabs

this help me to use multiple tabs with diferents sessions of guacamole..

sorry for my poor english, is not my main languaje.
